go/mysql/server.go

```diff
@@ -378,14 +378,24 @@ func (l *Listener) handle(conn net.Conn, connectionID uint32, acceptTime time.Ti
 
 	// See what auth method the AuthServer wants to use for that user.
 	negotiatedAuthMethod, err := negotiateAuthMethod(c, l.authServer, user, clientAuthMethod)
-	if err != nil {
-		// The client will disconnect if it doesn't understand
-		// the first auth method that we send, so we only have to send the
-		// first one that we allow for the user.
-		for _, m := range l.authServer.AuthMethods() {
-			if m.HandleUser(c, user) {
-				negotiatedAuthMethod = m
-				break
+
+	// We need to send down an additional packet if we either have no negotiated method
+	// at all or incomplete authentication data.
+	//
+	// The latter case happens for example for MySQL 8.0 clients until 8.0.25 who advertise
+	// support for caching_sha2_password by default but with no plugin data.
+	if err != nil || !negotiatedAuthMethod.HasCompleteAuthPluginData(clientAuthResponse) {
```

Comment:
I would just do a `len` check here.

Reply:
Doing the length check here now. We can always add back an abstraction if there's contextual knowledge needed.